llvm/docs/DTLTO.rst

```diff
@@ -11,45 +11,64 @@ DTLTO
+files, such as Bazel. DTLTO removes this requirement, allowing it to be used
+with any build process that supports in-process ThinLTO.
+
+The following commands show the steps used for the *separate thin-link*
+approach for a basic example:
+
+.. code-block:: console
+
+    1. clang -flto=thin -O2 file1.c file2.c -c
+    2. clang -flto=thin -O2 file1.o file2.o -fuse-ld=lld -Wl,--thinlto-index-only
+    3. clang -O2 -o file1.native.o -x ir file1.o -c -fthinlto-index=./file1.o.thinlto.bc
+    4. clang -O2 -o file2.native.o -x ir file2.o -c -fthinlto-index=./file2.o.thinlto.bc
+    5. clang file1.native.o file2.native.o -o a.out -fuse-ld=lld
 
-The current implementation of DTLTO has the following limitations:
+With DTLTO, steps 2.-5. are performed internally as part of the link step. The
+equivalent DTLTO commands for the above are:
 
-- The ThinLTO cache is not supported.
-- Only ELF and COFF platforms are supported.
-- Archives with bitcode members are not supported.
-- Only a very limited set of LTO configurations are currently supported, e.g.,
-  support for basic block sections is not currently available.
+.. code-block:: console
 
-Overview of Operation
----------------------
+    clang -flto=thin -O2 file1.c file2.c -c
+    clang -flto=thin -O2 file1.o file2.o -fuse-ld=lld -fthinlto-distributor=<distributor_process>
 
-For each ThinLTO backend compilation job, LLD:
+For DTLTO, LLD prepares the following for each ThinLTO backend compilation job:
 
-1. Generates the required summary index shard.
-2. Records a list of input and output files.
-3. Constructs a Clang command line to perform the ThinLTO backend compilation.
+- An individual index file and a list of input and output files (corresponds to
+  step 3. above).
+- A Clang command line to perform the ThinLTO backend compilations.
 
-This information is supplied, via a JSON file, to a distributor program that
-executes the backend compilations using a distribution system. Upon completion,
-LLD integrates the compiled native object files into the link process.
+This information is supplied, via a JSON file, to ``distributor_process``, which
+executes the backend compilations using a distribution system (corresponds to
+step 4. above). Upon completion, LLD integrates the compiled native object files
+into the link process and completes the link (corresponds to step 5. above).
 
-The design keeps the details of distribution systems out of the LLVM source
+This design keeps the details of distribution systems out of the LLVM source
 code.
 
+An example distributor that runs ``clang`` on the local system is included in
+the LLVM source tree. To run an example with that distributor, a command line
+such as the following can be used:
+
+.. code-block:: console
+
+   clang -flto=thin -fuse-ld=lld -O2 file1.o file2.o \
+     -fthinlto-distributor=$(which python3) \
+     -Xthinlto-distributor=$LLVMSRC/llvm/utils/dtlto/local.py
```

Comment:
I assume these don't exist yet as this PR is "LLVM only"? Shouldn't that second option have `-arg` at the end or is that the purpose of the `-X`, i.e. implies append?

Reply:
> I assume these don't exist yet as this PR is "LLVM only"?

Upstream reviewers seem quite relaxed about this as long as the Clang and LLD changes go in quickly once the LLVM stuff is agreed.

> Shouldn't that second option have `-arg` at the end or is that the purpose of the `-X`, i.e. implies append?

That's what the -X implies.